### ISSUE TITLE
Bunyan crashing production server for some reason

### ISSUE THREAD

**Issue by vrmerlin** [trentm/node-bunyan#390]
Our production server, which uses bunyan 1.8.0, has been periodically crashing hard lately. I found this error on the stdout when it crashed:

```
/www/node_modules/bunyan/lib/bunyan.js:1383
        throw new TypeError('cannot start a rotation when already rotating');
        ^

TypeError: cannot start a rotation when already rotating
    at RotatingFileStream.rotate (/www/node_modules/bunyan/lib/bunyan.js:1383:15)
    at null._onTimeout (/www/node_modules/bunyan/lib/bunyan.js:1268:28)
    at Timer.unrefdHandle (timers.js:312:14)
```

Here is how I'm setting up the logger:

```
    var logDirectory = '/app/runtime/speed/logs';
    fs.existsSync(logDirectory) || fs.mkdirSync(logDirectory);

    logger = bunyan.createLogger({
        name: "Shot planner Experiment Editor",
        streams: [
            {stream: process.stdout},
            {
                type: 'rotating-file',
                path: logDirectory + "/logfile.log",
                period: "1d",
                count: 5
            }
        ]
    });
```

Any ideas on what might be causing this? The server runs for several days without problems, before this happens.

Thanks,
John

**Comment by vrmerlin** [comment]
I was hoping I would get a reply -- as I said, this is impacting our production server. Any ideas on what might be going wrong here?

**Comment by voxpelli** [comment]
After giving the code a quick look, here's a quick guess until someone with better knowledge can take a look:

[RotatingFileStream.prototype._setRotationTimer()](https://github.com/trentm/node-bunyan/blob/master/lib/bunyan.js#L1258) doesn't cancel an existing timer before it sets a new one. That means that if for some reason some code would result in a new timer to be set up before an existing one is triggered, then two timers would be triggered and you could probably end up with the issue you're having.

Checking that file further it seems that the only two places in that file that call `_setRotationTimer()` are `_setupNextRot()` and `rotate()`, and the latter also calls the former.

Looking further it shows that the constructor actually calls both `rotate()` and `_setupNextRot()` if it rotates at the start (if `rotateAfterOpen` becomes true). That seems like a case that, if not handled properly could cause two rotations to be set up which would in turn cause the error that you see.

A way to at least mitigate the effects of whatever error you're running across could probably be by adding this to `_setRotationTimer()`, right before it sets the new `this.timeout` value:

```javascript
if (this.timeout) { clearTimeout(this.timeout); }
```

That way the error you run across would at least be much harder to trigger and I can see no harm caused by adding it after the quick look I just had at the code.

**Comment by vrmerlin** [comment]
Thanks, voxpelli. That sounds like the flavor of the problem. My build process is pulling from NPMJS though, I don't have a local version. Is there a patch possible that adds that?

**Comment by voxpelli** [comment]
@vrmerlin You can fork the repo here on GitHub, add the fix and temporarily change your dependency to the git URL of that repo, and if it works out then you could also use the branch on that repo to submit a PR and once that PR has been released you can move back your dependency to a npm version range.

**Comment by vrmerlin** [comment]
I've made the change on a fork, and will try it out in our next release.

**Comment by trentm** [comment]
@Rcomian ideas? suggestions?

**Comment by Rcomian** [comment]
I've not come across this particular one, sounds interesting. You might want to check out: https://www.npmjs.com/package/bunyan-rotating-file-stream

I've been running this under stress for several weeks now and I'm getting a bit more confidence in it. I'd be interested if it worked for you, @vmerlin

It handles that particular issue in a different way, since there are a few more triggers and concerns being dealt with. Hopefully it'll be a good fit.

**Comment by trentm** [comment]
I think this is the same bug as fixed by pull #386. That will be in version 1.8.1.
